clients/venice-thin-client/src/main/java/com/linkedin/venice/client/schema/RouterBackedSchemaReader.java

```diff
@@ -300,6 +290,294 @@ private SchemaEntry fetchSingleSchema(String requestPath, boolean isValueSchema)
+      }
+    } else {
+      if (hasSupersetSchema) {
+        latestSchemaId = supersetSchemaId;
+      } else {
+        latestSchemaId = maxSchemaId;
+      }
+    }
+    if (latestSchemaId == SchemaData.INVALID_VALUE_SCHEMA_ID) {
+      return;
+    }
+    // Perform latest value schema entry update.
+    if (latestValueSchemaEntry.get() == null || latestSchemaId == supersetSchemaId
+        || latestSchemaId > latestValueSchemaEntry.get().getId()) {
+      latestValueSchemaEntry.set(valueSchemaEntryMap.get(latestSchemaId));
+    }
+  }
+
+  private SchemaEntry maybeUpdateLatestValueSchemaEntry() {
+    SchemaEntry latest = latestValueSchemaEntry.get();
+    if (latest == null || shouldRefreshLatestValueSchemaEntry.get()) {
+      /**
+       * Every time it sees latestValueSchemaEntry is null or the flag to update latest schema entry is set to true,
+       * it will try to update it once.
+       * The update is expensive, but we expect it to be filled after the first fetch, as each store should have at least
+       * one active value schema.
+       */
+      updateAllValueSchemaEntriesAndLatestValueSchemaEntry(false);
+      shouldRefreshLatestValueSchemaEntry.compareAndSet(true, false);
+      latestValueSchemaEntry.set(latestValueSchemaEntry.get());
+      latest = latestValueSchemaEntry.get();
+    }
+    return latest;
+  }
+
+  /**
+   * This method will fetch the value schema by ID, and potentially will update the cached schema map.
+   * If the schema:
+   * (1) Exists in the cached schema map, we will directly return it.
+   * (2) Does not exist in the cached schema map, we will perform one time single schema fetch to request it from router,
+   * and update the result in the cached schema map.
+   * (3) If the schema was queried previously and not found, we will:
+   * -- (a) Return not found, if forceRefresh flag == false.
+   * -- (b) Perform one time router query, update and return the result if forceRefresh flag == true.
+   */
+  private SchemaEntry maybeUpdateAndFetchValueSchemaEntryById(int valueSchemaId, boolean forceRefresh) {
+    if (forceRefresh) {
+      SchemaEntry entry = fetchValueSchemaEntryFromRouter(valueSchemaId);
+      if (entry == null) {
+        valueSchemaEntryMap.put(valueSchemaId, NOT_EXIST_VALUE_SCHEMA_ENTRY);
+        return NOT_EXIST_VALUE_SCHEMA_ENTRY;
+      } else {
+        valueSchemaEntryMap.put(valueSchemaId, entry);
+        valueSchemaMapR.put(entry.getSchema(), valueSchemaId);
+        return entry;
+      }
+    } else {
+      return valueSchemaEntryMap.computeIfAbsent(valueSchemaId, (schemaId) -> {
+        SchemaEntry entry = fetchValueSchemaEntryFromRouter(valueSchemaId);
+        if (entry == null) {
+          return NOT_EXIST_VALUE_SCHEMA_ENTRY;
+        }
+        // Every time when we fetch a new value schema to cache during non-force-refresh logic, we should try to mark
+        // the flag as true.
+        shouldRefreshLatestValueSchemaEntry.compareAndSet(false, true);
```

Comment:
I don't fully get this variable, so can you explain more?

Reply:
When we fetch a new schema from backend, this means the latest value schema id we computed might be already stale, so next time when we are trying to get latest value schema ID, we should perform one time request to fetch it.